docs/Auth0/README.md

```diff
@@ -0,0 +1,237 @@
+```
+
+Just pick one of the listed suggestions and place it at the top of your `configure` function. If you want your data to persist across server reboots, use `config.prefer(FluentCache<SQLiteDatabase>.self, for: KeyedCache.self)`
+
+Imperial uses environment variables to access the client ID and secret to authenticate with Google. To allow Imperial to access these tokens, you will create these variables, called `AUTH0_DOMAIN`, `AUTH0_CLIENT_ID` and `AUTH0_CLIENT_SECRET`, with the domain, client ID, and secret assigned to them. Imperial can then access these vars and use there values to authenticate with Auth0.
+
+Now, all we need to do is register the Auth0 service in your main router method, like this:
+
+
+```swift
+try router.oAuth(from: Auth0.self, authenticate: "login", callback: "http://localhost/login/callback", scope: ["profile"]) { (request, token) in
+    print(token)
+    return Future(request.redirect(to: "/"))
+}
+```
+
+The `profile` scope will allow you to get more profile information about the user via the resulting access token.  Auth0 requires the `openid` scope, which will automatically be included for you.
+
+If you just want to redirect, without doing anything else in the callback, you can use the helper `Route.oAuth` method that takes in a redirect string:
+
+```swift
+try router.oAuth(from: Auth0.self, authenticate: "login", callback: "http://localhost/login/callback", redirect: "/")
+```
+
+The `callback` argument is the path you will go to when you want to authenticate the user. The `callback` argument has to be one of the URLs you entered in the Allowed Callback URLs on your Auth0 dashboard.
+
+The completion handler is fired when the callback route is called by the OAuth provider (Auth0). The access token is passed in and a response is returned.
+
+If you ever want to get the `access_token` in a route, you can use a helper method for the `Request` type that comes with Imperial:
+
+```swift
+let token = try request.accessToken()
+```
+
+Now that you are authenticating the user, you will want to protect certain routes to make sure the user is authenticated. You can do this by adding the `ImperialMiddleware` to a router group (or maybe your middleware config):
+
+```swift
+let protected = router.grouped(ImperialMiddleware())
+```
+
+Then, add your protected routes to the `protected` group:
+
+```swift
+protected.get("me", handler: me)
+```
+
+The `ImperialMiddleware` by default passes the errors it finds onto `ErrorMiddleware` where they are caught, but you can initialize it with a redirect path to go to if the user is not authenticated:
+
+```swift
+let protected = router.grouped(ImperialMiddleware(redirect: "/"))
+```
+
+## Supporting SSL
```

Comment:
@0xTim Unfortunately Auth0 requires the  Login URI to be https.  Perhaps I should add a comment about that.  If there's a better SSL getting started doc, I don't mind pointing to that.  I just didn't find one that was very straightforward for just getting up and running quickly.

Reply:
So looking at the docs, it doesn't explicitly state that https://auth0.com/docs/dashboard/reference/settings-application

It is confusing though! Have you tried to with `http://localhost:8080`? I suspect it would work. If so, we can remove this section

Reply:
Yeah, the documentation doesn't explicitly state it, other than saying it "usually takes the form of https://something".  But if you try http, it gives an error of: "Error! Payload validation error: 'Object didn't pass validation for format absolute-uri-or-empty: http://localhost:8080' on property initiate_login_uri (Initiate login uri, must be https and cannot contain a fragment)."

Reply:
Fair enough, most sites allow you to use HTTP for testing on localhost but I guess not Auth0! We'll keep it in then